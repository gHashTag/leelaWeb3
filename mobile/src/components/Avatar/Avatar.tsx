import React, { memo } from 'react'

import {
  ImageBackground,
  ImageStyle,
  Pressable,
  StyleProp,
  StyleSheet,
  View,
} from 'react-native'

import { useTheme } from '@react-navigation/native'
import { NeomorphView, Text, Loader } from 'components'
import { useTranslation } from 'react-i18next'
import { ms, s } from 'react-native-size-matters'

interface AvatarI {
  plan: number
  size: 'xLarge' | 'large' | 'medium' | 'small'
  avatar: string
  isAccept?: boolean
  additionalStyle?: StyleProp<ImageStyle>
  onPress?: () => void
  testID?: string
  showIcon?: boolean
  isLoading?: boolean
}

const Avatar = memo<AvatarI>(
  ({
    size = 'medium',
    plan,
    avatar,
    additionalStyle,
    onPress,
    testID = 'avatar',
    isLoading = false,
  }) => {
    const {
      colors: { background },
    } = useTheme()
    const { t } = useTranslation()

    const textPlan = plan < 10 ? `0${plan}` : `${plan}`

    let fontSize
    if (size === 'small') {
      fontSize = s(6)
    } else if (size === 'xLarge') {
      fontSize = s(15)
    } else {
      fontSize = s(10)
    }

    let badgeS
    if (size === 'small' || size === 'medium') {
      badgeS = styles.smallBadge
    } else {
      badgeS = styles.bigBadge
    }

    const newSize = [styles[size]]
    return (
      <Pressable onPress={onPress} testID={testID}>
        <NeomorphView viewStyle={{ ...styles.card, ...newSize[0] }}>
          <View style={styles.container}>
            {isLoading ? (
              <Loader />
            ) : avatar ? (
              <ImageBackground
                source={{ uri: avatar }}
                style={[styles[size], additionalStyle]}
                imageStyle={styles.container}
              >
                <View
                  style={[
                    styles.badge,
                    badgeS,
                    { backgroundColor: background },
                  ]}
                >
                  <Text
                    testID="avatar-title"
                    textStyle={[styles.plan, { fontSize }]}
                    title={textPlan}
                    h="h5"
                  />
                </View>
              </ImageBackground>
            ) : (
              <Text
                h={'h3'}
                testID="add-image-text"
                title={t('auth.addImage')}
                textStyle={styles.textStyle}
              />
            )}
          </View>
        </NeomorphView>
      </Pressable>
    )
  },
)

const styles = StyleSheet.create({
  badge: {
    alignItems: 'center',
    borderRadius: s(180),
    flexDirection: 'row',
    justifyContent: 'center',
    padding: s(4),
    position: 'absolute',
  },
  bigBadge: {
    alignSelf: 'center',
    bottom: s(-15),
    height: s(30),
    width: s(30),
  },
  card: {
    alignItems: 'center',
    borderRadius: s(60),
    height: s(65),
    justifyContent: 'center',
    width: s(65),
  },
  container: {
    alignItems: 'center',
    borderRadius: ms(130),
    justifyContent: 'center',
  },
  large: {
    height: s(70),
    width: s(70),
  },
  medium: {
    height: ms(55, 0.9),
    width: ms(55, 0.9),
  },
  plan: {
    left: 0.5,
    top: 1,
  },
  small: {
    height: s(36),
    width: s(36),
  },
  smallBadge: {
    alignSelf: 'center',
    bottom: s(-8),
  },
  textStyle: {
    alignSelf: 'center',
    lineHeight: 23,
    textAlign: 'center',
    top: 5,
    width: 110,
  },
  xLarge: {
    borderRadius: s(70),
    height: ms(134),
    justifyContent: 'center',
    marginLeft: 1,
    width: ms(134),
  },
})

export { Avatar }
