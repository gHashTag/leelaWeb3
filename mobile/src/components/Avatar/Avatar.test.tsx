import React from 'react'

import { fireEvent, render } from '@testing-library/react-native'
import renderer from 'react-test-renderer'

import '@testing-library/jest-native/extend-expect'
import { Avatar } from './Avatar'

describe('Avatar Component', () => {
  it('matches the snapshot', () => {
    const tree = renderer
      .create(
        <Avatar
          plan={5}
          size="medium"
          avatar="avatar-url"
          testID="avatar-component"
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with default props', () => {
    const { getByTestId } = render(
      <Avatar
        plan={5}
        size="medium"
        avatar="avatar-url"
        testID="avatar-component"
      />,
    )
    expect(getByTestId('avatar-component')).toBeTruthy()
  })

  it('calls onPress function when pressed', () => {
    const onPressMock = jest.fn()
    const { getByTestId } = render(
      <Avatar
        plan={5}
        size="medium"
        avatar="avatar-url"
        onPress={onPressMock}
        testID="avatar-component"
      />,
    )
    const avatarComponent = getByTestId('avatar-component')
    fireEvent.press(avatarComponent)
    expect(onPressMock).toHaveBeenCalled()
  })

  it('displays time icon when isAccept is false', () => {
    const { queryByTestId } = render(
      <Avatar
        plan={5}
        size="medium"
        avatar="avatar-url"
        isAccept={false}
        testID="avatar-component"
      />,
    )
    expect(queryByTestId('avatar-title')).toBeNull()
    expect(queryByTestId('time-icon')).toBeTruthy()
  })

  it('displays text title when isAccept is true', () => {
    const { queryByTestId } = render(
      <Avatar
        plan={5}
        size="medium"
        avatar="avatar-url"
        isAccept={true}
        testID="avatar-component"
      />,
    )
    expect(queryByTestId('avatar-title')).toBeTruthy()
    expect(queryByTestId('avatar-title')).toHaveTextContent('05')
  })
})
