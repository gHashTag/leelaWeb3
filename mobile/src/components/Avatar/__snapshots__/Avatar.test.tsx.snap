// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Component matches the snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="avatar-component"
>
  <View
    style={
      {
        "alignItems": "center",
        "borderRadius": 204.28571428571428,
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        [
          {
            "height": 101.42857142857142,
            "width": 101.42857142857142,
          },
          undefined,
        ]
      }
    >
      <Image
        source={
          {
            "uri": "avatar-url",
          }
        }
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "height": 101.42857142857142,
              "width": 101.42857142857142,
            },
            {
              "alignItems": "center",
              "borderRadius": 204.28571428571428,
              "justifyContent": "center",
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 385.7142857142857,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8.571428571428571,
              "position": "absolute",
            },
            {
              "bottom": 2.142857142857143,
              "right": -2.142857142857143,
            },
            {
              "backgroundColor": "rgb(242, 242, 242)",
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#FFB700",
                "fontSize": 32.14285714285714,
              },
              undefined,
              {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
          testID="time-icon"
        >
          î» 
        </Text>
      </View>
    </View>
  </View>
</View>
`;
